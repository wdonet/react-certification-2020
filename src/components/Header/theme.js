const base = {
  easeOutBack: 'cubic-bezier(0.34, 1.56, 0.64, 1)',
  colorWhite: 'rgb(255, 255, 255)',
  colorBlack: 'rgb(0, 0, 0)',
};

const dark = {
  id: 'dark',
  ...base,
  backgroundColor: '#333',
  textColor: 'black',
};

const light = {
  id: 'light',
  ...base,
  backgroundColor: '#666',
  textColor: 'white',
};

export const theme = { dark, light };
