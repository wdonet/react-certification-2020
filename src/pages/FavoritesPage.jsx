import React from 'react';
import { StoreContext } from '../contexts/Store'
import Content from '../components/Content';
import VideoCard from '../components/VideoCard';
import styled from 'styled-components';
import {
  Link
} from "react-router-dom";


const StyledRemoveFavorite =styled.button`
    height: 10px;
    height: 50px;
    margin: 5px;
    padding:5px;
`



const StyledLink = styled(Link)`
  color: ${(props) => props.theme.text};
  text-decoration: none;
  &:focus, &:hover, &:visited, &:link, &:active {
      text-decoration: none;
  }
  margin-top: 16px;
`


const FavoritesPage = () => {

  const data = {items:[]}
  const {
    "favorites": [favorites, setFavorites],
  } = React.useContext(StoreContext)

  for (const [key, value] of Object.entries(favorites)) {
      console.log(key)
      data.items.push(value);
  }


  const removeFav = (videoid) => {
    const clone = JSON.parse(JSON.stringify(favorites));
    delete clone[videoid];
    setFavorites(clone);
  }
  return (
    <Content title="Favorites"  is>{(data.items && data.items.length>0) ?
      data.items.map((video) => {
        return (
          <>
            <StyledLink key={video.id.videoId} to={`/video/${video.id.videoId}?fav=true`}>
              <VideoCard
                key={video.id.videoId} 
                videoId={video.id.videoId}
                title={video.snippet.title}
                description={video.snippet.description}
                image={video.snippet.thumbnails.high.url}
              />
            </StyledLink>
            <StyledRemoveFavorite onClick={ ()=>removeFav(video.id.videoId) }>Remove</StyledRemoveFavorite>
          </>
        )
      })
    :"No favorites, go Home and start adding some!"}
    </Content>
  )
}

export default FavoritesPage;
