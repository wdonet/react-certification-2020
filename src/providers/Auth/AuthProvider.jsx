import React, { useState, useEffect, useCallback } from 'react';

import { AUTH_STORAGE_KEY } from '../../utils/constants';
import { loginUser } from '../../mocks/loginApi';
import { storage } from '../../utils/storage';

const AuthContext = React.createContext(null);

function AuthProvider({ authenticatedJest, children }) {
  const [authenticated, setAuthenticated] = useState(authenticatedJest);

  useEffect(() => {
    const lastAuthState = storage.get(AUTH_STORAGE_KEY);
    const isAuthenticated = Boolean(lastAuthState);

    setAuthenticated(isAuthenticated || authenticatedJest);
  }, []);

  const login = async ({ username, password }) => {
    try {
      const authorized = await loginUser({ username, password });
      if (authorized) {
        setAuthenticated(true);
        storage.set(AUTH_STORAGE_KEY, true);
      }
      return { success: true };
    } catch (error) {
      return { error };
    }
  };

  const logout = useCallback(() => {
    setAuthenticated(false);
    storage.set(AUTH_STORAGE_KEY, false);
  }, []);

  return (
    <AuthContext.Provider value={{ login, logout, authenticated }}>
      {children}
    </AuthContext.Provider>
  );
}

export { AuthContext };
export default AuthProvider;
