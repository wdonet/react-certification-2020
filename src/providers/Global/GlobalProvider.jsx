import React, { useReducer } from 'react';
import reducer from './reducer';
import { saveFavorite, deleteFavorite, setTheme, setSelectedVideo } from './actions';

const GlobalContext = React.createContext(null);

const GlobalProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    favorites: {},
    theme: 'light',
    selectedVideo: '',
  });

  const { favorites, theme, selectedVideo } = state;

  return (
    <GlobalContext.Provider
      value={{
        favorites,
        selectedVideo,
        setSelectedVideo: (videoId) => setSelectedVideo(dispatch, videoId),
        saveFavorite: (videoId, videoDetails) =>
          saveFavorite(dispatch, videoId, videoDetails),
        deleteFavorite: (videoId) => deleteFavorite(dispatch, videoId),
        theme,
        setTheme: (newTheme) => setTheme(dispatch, newTheme),
      }}
    >
      {children}
    </GlobalContext.Provider>
  );
};

GlobalProvider.propTypes = {};

export { GlobalContext };
export default GlobalProvider;
