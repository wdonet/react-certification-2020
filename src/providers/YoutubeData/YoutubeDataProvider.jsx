import React, { useEffect, useReducer } from 'react';
import { getVideosByQuery, getVideoDetails, getRelatedVideos } from './api';
import reducer from './reducer';
import { setSearchTerm, setVideos } from './actions';

const YoutubeDataContext = React.createContext(null);

const YoutubeDataProvider = ({ iframeAPIReady, children }) => {
  const [state, dispatch] = useReducer(reducer, {
    videos: [],
    searchTerm: '',
  });

  const { videos, searchTerm } = state;

  const fetchVideos = async (query) => {
    try {
      const newVideos = await getVideosByQuery(query);
      setVideos(dispatch, newVideos);
    } catch (error) {
      console.error(error);
    }
  };

  const fetchVideoDetails = async (videoId) => {
    try {
      const videoDetails = await getVideoDetails(videoId);
      return videoDetails;
    } catch (error) {
      console.error(error);
    }
  };

  const fetchRelatedVideos = async (videoId) => {
    try {
      const relatedVideos = await getRelatedVideos(videoId);
      return relatedVideos;
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    fetchVideos();
  }, []);

  return (
    <YoutubeDataContext.Provider
      value={{
        videos,
        fetchVideos,
        iframeAPIReady,
        fetchVideoDetails,
        fetchRelatedVideos,
        searchTerm,
        setSearchTerm: (search) => setSearchTerm(dispatch, search),
      }}
    >
      {children}
    </YoutubeDataContext.Provider>
  );
};

YoutubeDataProvider.propTypes = {};

export { YoutubeDataContext };
export default YoutubeDataProvider;
