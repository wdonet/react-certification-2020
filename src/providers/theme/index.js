export { default, useTheme } from './Theme.provider';
