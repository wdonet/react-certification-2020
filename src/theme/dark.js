const dark = {
  palette: {
    primary: {
      main: '#556cd6',
      transparent: '3,45,70,0.4',
    },
    secondary: {
      main: '#303030',
      light: '#424242',
    },
    accent: '#19857b',
  },
  typography: {
    title: {
      color: '#fff',
    },
    text: {
      color: '#fafafa',
    },
  },
  appBar: {
    height: {
      desktop: '64px',
    },
  },
  profile: {
    primary: '#303030',
    secondary: '#636363',
  },
};

export default dark;
