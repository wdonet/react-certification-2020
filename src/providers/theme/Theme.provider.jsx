// React
import React, { useState, useEffect, useContext, useCallback } from 'react';
// Utils
import { THEME_SELECTED } from '../../utils/constants';
import { storage } from '../../utils/storage';
// Themes
import light from '../../theme/light';
import dark from '../../theme/dark';

const ThemeContext = React.createContext({ theme: light, selected: false });

const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error(`Can't use "useTheme" without an ThemeProvider!`);
  }
  return context;
};

const ThemeBaseProvider = ({ children }) => {
  const [theme, setTheme] = useState(light);
  const [selected, setSelected] = useState(false);

  useEffect(() => {
    const lastTheme = storage.get(THEME_SELECTED);
    const whichTheme = lastTheme ? dark : light;
    setSelected(lastTheme);
    setTheme(whichTheme);
  }, []);

  const toggleTheme = useCallback((toggle) => {
    setTheme(toggle ? dark : light);
    storage.set(THEME_SELECTED, toggle);
  }, []);

  return (
    <ThemeContext.Provider value={{ theme, selected, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export { useTheme, ThemeContext };
export default ThemeBaseProvider;
