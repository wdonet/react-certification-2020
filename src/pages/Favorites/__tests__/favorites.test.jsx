import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import '@testing-library/jest-dom/extend-expect';
import 'jest-styled-components';

import FavoritesPage from '../index';
import AuthProvider from '../../../providers/Auth';
import YoutubeDataProvider from '../../../providers/YoutubeData';
import GlobalProvider from '../../../providers/Global';

import { lightTheme } from '../../../theme';

describe('Favorites', () => {
  test('renders correctly', () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <GlobalProvider>
            <AuthProvider>
              <YoutubeDataProvider>
                <ThemeProvider theme={lightTheme}>
                  <FavoritesPage />
                </ThemeProvider>
              </YoutubeDataProvider>
            </AuthProvider>
          </GlobalProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
