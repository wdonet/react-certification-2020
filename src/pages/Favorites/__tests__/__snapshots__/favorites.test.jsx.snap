// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favorites renders correctly 1`] = `
.c0 {
  text-align: center;
}

.c1 {
  color: #333333;
}

<section
  className="c0"
>
  <h2
    className="c1"
  >
    Favorites
  </h2>
  No favorites added yet
</section>
`;
